From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CTS Exam Bank <exam@example.com>
Date: Mon, 1 Jan 2024 00:00:00 +0000
Subject: [PATCH] [BUG] DisplayInfo: swap ownerPackageName and uniqueId order in writeToParcel

This patch introduces a subtle bug in DisplayInfo.writeToParcel() where
the order of ownerPackageName and uniqueId fields is swapped compared to
readFromParcel(). This causes the two fields to have their values
exchanged when DisplayInfo is transmitted across process boundaries
via Binder IPC.

Bug Type: Parcelable serialization order mismatch
Affected Files:
  1. DisplayInfo.java - writeToParcel order differs from readFromParcel
  2. LogicalDisplay.java - sets both fields correctly
  3. DisplayManagerGlobal.java - receives corrupted data via IPC

Impact:
- VirtualDisplay's uniqueId becomes the owner package name
- VirtualDisplay's ownerPackageName becomes the uniqueId
- Display identification and ownership checks fail

CTS Tests that will fail:
- android.display.cts.VirtualDisplayTest#testPrivateVirtualDisplay
- android.display.cts.VirtualDisplayTest#testPublicPresentationVirtualDisplay

---
 core/java/android/view/DisplayInfo.java | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/core/java/android/view/DisplayInfo.java b/core/java/android/view/DisplayInfo.java
index 1234567890a..1234567890b 100644
--- a/core/java/android/view/DisplayInfo.java
+++ b/core/java/android/view/DisplayInfo.java
@@ -626,8 +626,8 @@ public final class DisplayInfo implements Parcelable {
         dest.writeInt(state);
         dest.writeInt(committedState);
         dest.writeInt(ownerUid);
-        dest.writeString8(ownerPackageName);
-        dest.writeString8(uniqueId);
+        dest.writeString8(uniqueId);          // BUG: should be ownerPackageName
+        dest.writeString8(ownerPackageName);  // BUG: should be uniqueId
         dest.writeInt(removeMode);
         dest.writeFloat(refreshRateOverride);
         dest.writeFloat(brightnessMinimum);
-- 
2.34.1
